Library/Homebrew/system_command.rb

```diff
@@ -276,9 +273,56 @@ def each_output_line(&block)
   rescue Interrupt
     Process.kill("INT", raw_wait_thr.pid) if raw_wait_thr && !sudo?
     raise Interrupt
-  rescue SystemCallError => e
-    @status = $CHILD_STATUS
-    @output << [:stderr, e.message]
+  ensure
+    raw_stdin&.close
+    raw_stdout&.close
+    raw_stderr&.close
+  end
+
+  sig {
+    params(
+      env:        T::Hash[String, String],
+      executable: String,
+      args:       String,
+      options:    T.untyped,
+    ).returns([IO, IO, IO, Thread])
+  }
+  def exec3(env, executable, *args, **options)
+    in_r, in_w = IO.pipe
+    options[:in] = in_r
+    in_w.sync = true
+
+    out_r, out_w = IO.pipe
+    options[:out] = out_w
+
+    err_r, err_w = IO.pipe
+    options[:err] = err_w
+
+    pid = fork do
+      Process::UID.change_privilege(Process.euid) if reset_uid? && Process.euid != Process.uid
+
+      exec(
+        env.merge({ "COLUMNS" => Tty.width.to_s }),
+        [executable, executable],
+        *args,
+        **options,
```

Comment:
Do `Open.popen3` and `Kernel.exec` take the same `options`?

Reply:
Yes:

https://docs.ruby-lang.org/en/3.3/Open3.html
https://docs.ruby-lang.org/en/3.3/Kernel.html#method-i-exec

Internally Open3 uses Process.spawn which itself is a fancy fork + exec.